pkg/controllers/build/credentials_controller.go

```diff
@@ -0,0 +1,226 @@
+}
+
+func (r *CredentialReconciler) reconcile(ctx context.Context, log logr.Logger, serviceAccount *corev1.ServiceAccount, namespace string) (ctrl.Result, error) {
+	secretNames := sets.NewString()
+	var secrets corev1.SecretList
+	if err := r.List(ctx, &secrets, client.InNamespace(namespace), MatchingLabels(buildv1alpha1.CredentialLabelKey)); err != nil {
+		log.Error(err, "Failed to get Secrets", "serviceaccount", serviceAccount)
+		return ctrl.Result{Requeue: true}, err
+	}
+	for _, secret := range secrets.Items {
+		secretNames.Insert(secret.Name)
+	}
+
+	if serviceAccount.Name == "" {
+		if needed, err := r.isServiceAccountNeeded(ctx, secretNames, namespace); err != nil {
+			return ctrl.Result{}, err
+		} else if needed {
+			serviceAccount, err := r.createServiceAccount(ctx, secretNames, namespace)
+			if err != nil {
+				log.Error(err, "Failed to create ServiceAccount", "serviceaccount", serviceAccount)
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		serviceAccount, err := r.reconcileServiceAccount(ctx, serviceAccount, secretNames)
+		if err != nil {
+			log.Error(err, "Failed to reconcile ServiceAccount", "serviceaccount", serviceAccount)
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *CredentialReconciler) reconcileServiceAccount(ctx context.Context, existingServiceAccount *corev1.ServiceAccount, desiredBoundSecrets sets.String) (*corev1.ServiceAccount, error) {
+	serviceAccount := existingServiceAccount.DeepCopy()
+	boundSecrets := sets.NewString(strings.Split(serviceAccount.Annotations[buildv1alpha1.CredentialsAnnotationKey], ",")...)
+	removeSecrets := boundSecrets.Difference(desiredBoundSecrets)
+
+	secrets := []corev1.ObjectReference{}
+	// filter out secrets no longer bound
+	for _, secret := range serviceAccount.Secrets {
+		if !removeSecrets.Has(secret.Name) {
+			secrets = append(secrets, secret)
+		}
+	}
+	// add new secrets
+	for _, secret := range desiredBoundSecrets.Difference(boundSecrets).UnsortedList() {
+		secrets = append(secrets, corev1.ObjectReference{Name: secret})
+	}
+	serviceAccount.Secrets = secrets
+
+	if serviceAccount.Annotations == nil {
+		serviceAccount.Annotations = map[string]string{}
+	}
+	serviceAccount.Annotations[buildv1alpha1.CredentialsAnnotationKey] = strings.Join(desiredBoundSecrets.List(), ",")
+
+	if serviceAccountSemanticEquals(serviceAccount, existingServiceAccount) {
+		// No differences to reconcile.
+		return serviceAccount, nil
+	}
+
+	return serviceAccount, r.Update(ctx, serviceAccount)
+}
+
+func (r *CredentialReconciler) isServiceAccountNeeded(ctx context.Context, secretNames sets.String, namespace string) (bool, error) {
+	if secretNames.Len() != 0 {
+		return true, nil
+	}
+	var applications buildv1alpha1.ApplicationList
+	if err := r.List(ctx, &applications, client.InNamespace(namespace)); err != nil {
+		return false, err
+	} else if len(applications.Items) != 0 {
+		return true, nil
+	}
+	var functions buildv1alpha1.FunctionList
```

Comment:
should we check for containers as well?

Reply:
Containers don't produce build pods, so they don't need a service account